### ISSUE TITLE
Ember Inspector not detecting electron app

### ISSUE THREAD

**Issue by eg271** [adopted-ember-addons/ember-electron#532]
Using Ember 3.19.0
Node v14.4.0
Windows 10

Perhaps related to Ember Inspector's "Allow access to file URLs" option (is this enabled automatically by ember-electron)?

```
ember new electron-test
cd electron-test
ember install ember-electron@3.0.0-beta.3
ember electron
```
...
```
Starting Electron...

Installing Devtron from C:\Users\eg\Projects\electron-test\electron-app\node_modules\devtron
Failed to install Devtron:  TypeError: electron.BrowserWindow.addDevToolsExtension is not a function
    at Object.exports.install (C:\Users\eg\Projects\electron-test\electron-app\node_modules\devtron\api.js:13:35)
    at App.<anonymous> (C:\Users\eg\Projects\electron-test\electron-app\src\index.js:32:26)
    at App.emit (events.js:223:5)
(node:140720) ExtensionLoadWarning: Warnings loading extension at C:\Users\eg\AppData\Roaming\electron-test\extensions\bmdblncegkenkacieihfhpjfppoconhi: Unrecognized manifest key 'page_action'. Unrecognized manifest key 'update_url'. Permission 'contextMenus' is unknown or URL pattern is malformed. Cannot load extension with file or directory name _metadata. Filenames starting with "_" are reserved for use by the system.
```

Go to Ember tab in Developer Tools in the electron app window.  -> "Ember application not detected!"

But,
npm start
Navigate to localhost:4200 and go to Ember tab.  Works ok.

**Comment by rwwagner90** [comment]
This is definitely a known issue. I have not had a chance to debug it further yet.

**Comment by dtolios** [comment]
Just ran into this issue too. Is there any advice on where the bug might be?

**Comment by rwwagner90** [comment]
@dtolios I would guess the comment above would be a good place to start looking:

> Perhaps related to Ember Inspector's "Allow access to file URLs" option (is this enabled automatically by ember-electron)?

**Comment by dtolios** [comment]
Hmm is that similar to the sourcemap fix that went into the 3.0.0-beta branch where electron-protocol-serve was remove in favor of file:// ?

**Comment by rwwagner90** [comment]
@dtolios I don't think so. Ember Inspector needs a flag set to work with file urls, so we need a way to set that in Electron.

**Comment by bendemboski** [comment]
https://github.com/electron/electron/issues/24011 😢 

@dtolios it might be related to the same root cause -- Chromium reimplemented how they load assets, including local filesystem assets. The problem with source maps was that one effect of this Chromium change was the requests sent by the dev tools to load assets like source maps stopped going through the route that Electron's protocol handlers can intercept, so we could no longer supply source maps over a custom protocol. It's possible that this problem is related, but as you can see from that linked issue above, unfortunately it's not due to some unusual thing we are doing that we can just stop doing (like `electron-protocol-serve`), but due to some breaking changes in Chromium that need to be accounted for in Electron before anybody can use extensions.

It's not a great solution, but you could try running Electron 8 -- I believe the `ember-electron` 3.0 beta supports it, or if not, it's probably just some API calls in the blueprint-generated files that would need to be updated to use the older Electron APIs.

**Comment by dtolios** [comment]
Hey thanks for the prompt responses! Yeah I'll give electron 8 a try and if that doesn't work I can live without the inspector for now. 

I'll keep my eyes out for when 3.0.0 goes to full release as well! Thanks again and great work on this project, it is really awesome to have :)

**Comment by rwwagner90** [comment]
Looks like file urls don't work for extensions in Electron, so this is blocked by https://github.com/electron/electron/issues/24011

**Comment by rwwagner90** [comment]
@evoactivity I think we need to wait for Electron to support enabling file protocol for dev tools. There is a PR out for it, so hopefully we'll get some movement on this soon.

**Comment by bendemboski** [comment]
Yeah, as we learned through the whole migration from the serve protocol to the file protocol during this beta, file: vs. not-file: protocol has quite a lot of implications, and I think it would probably be fairly difficult to make our tooling cleanly support both, and from a software engineering methodology standpoint I think it would be a big mistake to have our dev environment differ that much from our production (and testing) environments.

I could potentially see building the instrumentation into the core `ember-cli` commands to allow applications to opt in to loading the app over HTTP, but I still think it would take quite a bit of effort and introduce significant additional maintenance overhead to this addon, and doesn't seem like good ROI to me.

**Comment by rwwagner90** [comment]
This is now supported in Electron 11.3.0! https://github.com/electron/electron/releases/tag/v11.3.0

**Comment by evoactivity** [comment]
I've had time to get back on my electron app today and I thought I'd outline the steps I needed to take to get emeber inspector to work.

1. In `electron-app/package.json` upgrade electron dependancy to 11.3.0 (12 crashed when trying to open electron, not sure why yet)
2. In `electron-app/package.json` upgrade electron-devtools-installer dependancy to latest github master (git://github.com/MarshallOfSound/electron-devtools-installer.git)
3. Update the call to `installExtension` in `electron-app/index.js` to use the new allowFileAccess option
```javascript
await installExtension(EMBER_INSPECTOR, {
  loadExtensionOptions: { allowFileAccess: true },
  forceDownload: true,
});
```

**Comment by rwwagner90** [comment]
@evoactivity is `forceDownload` required? I don't think I needed that, but not sure. Would you like to open a PR to update our blueprints for `installExtension`? It's used here https://github.com/adopted-ember-addons/ember-electron/blob/75ddbec7085aa0a210e20e9e569558b910ed2fa2/forge/files/tests/index.js#L22 and here https://github.com/adopted-ember-addons/ember-electron/blob/75ddbec7085aa0a210e20e9e569558b910ed2fa2/forge/files/src/index.js#L42

**Comment by evoactivity** [comment]
@rwwagner90 happy to make the change, one question though, since this relies on using an unpublished version of electron-devtools-installer should we hold off until a new release is published to npm?

**Comment by rwwagner90** [comment]
@evoactivity I think we are fine to go ahead and make the change. For older versions, it would be a noop.

**Comment by knownasilya** [comment]
How did you use the master version of https://github.com/MarshallOfSound/electron-devtools-installer since there is no `dist/` directory since it's not precompiled?  Anyone have a fork with the dist?

**Comment by bendemboski** [comment]
> How did you use the master version of https://github.com/MarshallOfSound/electron-devtools-installer since there is no dist/ directory since it's not precompiled? Anyone have a fork with the dist?

since that comment was posted, a new release of `electron-devtools-installer` was published, and `v3.2.0` currently points to the head of master (no commits since it was released), so you should be able to use that.

**Comment by knownasilya** [comment]
Yeah, just realized that. Seems like I was already using that. The only change I needed was:

```js
   await installExtension(EMBER_INSPECTOR, {
        loadExtensionOptions: { allowFileAccess: true },
      });
```
